### ISSUE TITLE
More array functions for models / collections

### ISSUE THREAD

**Issue by NonPolynomial** [miragejs/miragejs#306]
The collection already has `filter`, `slice` and `sort`.
But it would be nice to also have `map`,  `forEach` (could also be `each`) and `reduce`.

**Comment by samselikoff** [comment]
Yes – we actually used to do this but what makes it hard is that those methods don't necessarily return Collections (because you can return anything from them). A Collection needs a `modelName` property on it so it can be properly serialized.

It was also causing issues because folks thought collections were real JS arrays and they were trying to use Lodash methods on them and they were breaking in unexpected ways.

I'm not sure what I would do today + whether there's now a way to truly actually subclass arrays in JS.

Today the escape hatch is to do `collection.models` to access the underlying JavaScript Array of models and go to town on it.

**Comment by NonPolynomial** [comment]
> Today the escape hatch is to do collection.models to access the underlying JavaScript Array of models and go to town on it.

thats what I currently use at the moment

> It was also causing issues because folks thought collections were real JS arrays and they were trying to use Lodash methods on them and they were breaking in unexpected ways.

and using alternative function names?
e.g. `forEach` -> `each`

**Comment by samselikoff** [comment]
Yeah we could I'm just hesitant because it feels like extra APIs to support, it's something new to learn, and folks could still get confused and try to use libs like lodash on them. `.models` or `toArray()` feels like a decent middleground.

Could you share with me your use case (the last time you used one of those methods)?

**Comment by NonPolynomial** [comment]
I did a prototype for an App-Store, just to refresh some Angular basics.
So i had the models `App` and `Rating`
and the relation `App --|1:n|-> Rating`
I read in the documentation that `association` just works with `belongTo` but not with `hasMany` and I wanted to create random ratings for my apps when I call `server.createList('app', 4)` and to add a calculated field of the total rating, that the "frontend" doesn't have to calculate it itself.

If I had such array function it would be a little easier to achieve.

Or I totally missed something about calculated fields in the docs 😁

**Comment by samselikoff** [comment]
I would use [the afterCreate factory hook](https://miragejs.com/docs/main-concepts/factories#the-aftercreate-hook) along with a trait for this:

```js
import { Server, Model, Factory, trait } from 'miragejs';

new Server({
  models: {
    app: Model.extend({
      ratings: hasMany()
    }),
    rating: Model
  },

  factories: {
    app: {
      withRatings: trait({
        afterCreate(app, server) {
          app.update({
            ratings: server.createList('rating', 3, { app })
          })
        }
      })
    }
  },

  seeds(server) {
    server.createList('app', 4, 'withRatings') // boom, 4 apps with ratings
  }
})
```

Of course that `afterCreate` hook can do anything, make a random number of ratings, use other traits on the `rating` factory and so on.

**Comment by NonPolynomial** [comment]
I know, but the problem wasn't to create the related ratings but to add a calculated field in each app for the total rating

e.g.
App
- ratings:
  - rating: 5 stars
  - rating: 4 stars
  - rating 3 stars
- totalRating: 4 stars

those array functions would make it more easy to iterate over the ratings of an app to calculate such a field

**Comment by samselikoff** [comment]
Ahh I see. Yes, unfortunately we don't have good support for this right now, it's definitely a gap in Mirage that will be closed (sooner rather than later because people keep running into it).

Does it have to be on the model or just the response? If response, I would use the Serializer layer to add computed/derived fields. Most likely in the `serialize()` hook https://miragejs.com/api/classes/serializer/#serialize

**Comment by samselikoff** [comment]
Related https://github.com/miragejs/discuss/issues/45
